src/chainstate/stacks/index/storage.rs

```diff
@@ -747,85 +840,89 @@ impl TrieFileStorage {
-                        h
-                    },
-                    Err(e) => {
-                        let h = self.read_tmp_block_root_hash(bhh)?;
-                        trace!("Read {:?} from tmp file for {:?} instead", &h, bhh);
-                        h
-                    }
-                };
-                ret.insert(root_hash, bhh.clone());
             }
+
+            if *bhh == TrieFileStorage::block_sentinel() {
+                continue;
+            }
+
+            let root_hash = match self.read_block_root_hash(bhh) {
+                Ok(h) => {
+                    h
+                },
+                Err(e) => {
+                    let h = self.read_tmp_block_root_hash(bhh)?;
+                    trace!("Read {:?} from tmp file for {:?} instead", &h, bhh);
+                    h
+                }
+            };
+
+            ret.insert(root_hash.clone(), bhh.clone());
         }
 
-        let (last_extended_opt, last_extended_trie_opt) = match (self.last_extended.take(), self.last_extended_trie.take()) {
-            (Some(bhh), Some(mut trie_ram)) => {
+        let last_extended = match self.last_extended.take() {
+            Some((bhh, mut trie_ram)) => {
                 let ptr = TriePtr::new(set_backptr(TrieNodeID::Node256), 0, 0);
 
-                let mut root_hash_bytes = Vec::with_capacity(TRIEHASH_ENCODED_SIZE);
-                trie_ram.read_node_hash_bytes(&ptr, &mut root_hash_bytes)?;
+                let root_hash = trie_ram.read_node_hash(&ptr)?;
 
-                // safe because this is TRIEHASH_ENCODED_SIZE bytes long
-                let root_hash = trie_hash_from_bytes(&root_hash_bytes);
-                
-                ret.insert(root_hash, bhh.clone());
-                (Some(bhh), Some(trie_ram))
-            }
-            (_, _) => {
-                (None, None)
+                ret.insert(root_hash.clone(), bhh.clone());
+                Some((bhh, trie_ram))
+            },
+            _ => {
+                None
             }
         };
 
-        self.last_extended = last_extended_opt;
-        self.last_extended_trie = last_extended_trie_opt;
+        self.last_extended = last_extended;
 
         Ok(ret)
     }
 
     /// Extend the forest of Tries to include a new block.
     pub fn extend_to_block(&mut self, bhh: &BlockHeaderHash) -> Result<(), Error> {
-        if self.fork_table.size() > 0 {
-            if !self.fork_table.contains(&self.cur_block) {
-                return Err(Error::CorruptionError(format!("Current block {:?} not in fork table", &self.cur_block)));
-            }
-        }
-        if self.fork_table.contains(bhh) {
-            trace!("Block {:?} is in the fork table already", bhh);
-            return Err(Error::ExistsError);
-        }
+        /*
+
+        TODO:
+          must still test that:
+             * cur_block has an identifier.
+             * cur_block has written it's block height
+        
+         */
```

Comment:
Are these tests written now?

Reply:
No -- I'm now pretty sure that they should not be tested/asserted here.

The first property isn't a required assertion here (and would probably break `trie.rs` unit tests). The block identifier is _only_ ever set on a node copy forward. The block identifier is checked then, and (as of d31f0207) throws a corruption error if it is not set.

The second assertion is checked and throws a corruption error in `MARF::begin`.